feature/gribi/otg_tests/hierarchical_weight_resolution_test/hierarchical_weight_resolution_test.go

```diff
@@ -0,0 +1,739 @@
+		gribiEntries = append(gribiEntries, nh)
+		if i == 0 {
+			nextHopWeights = append(nextHopWeights, nhInfo{index: nhIdx, weight: 1})
+		} else {
+			nextHopWeights = append(nextHopWeights, nhInfo{index: nhIdx, weight: 16})
+		}
+		nhIdx++
+	}
+	nhg3 := nextHopGroupEntry(3, defaultVRF, nextHopWeights)
+	ipEntry3 := ipv4Entry(nhgIPv4EntryMap[3], defaultVRF, 3, defaultVRF)
+	gribiEntries = append(gribiEntries, nhg3, ipEntry3)
+
+	gRIBI.Modify().AddEntry(t, gribiEntries...)
+
+	// Set up NH#1, NH#2, NHG#1, IPv4Entry(198.18.196.1/22).
+	nh1 := nextHopEntry(1, defaultVRF, nhEntryIP1)
+	nh2 := nextHopEntry(2, defaultVRF, nhEntryIP2)
+	nhg1 := nextHopGroupEntry(1, defaultVRF, []nhInfo{{index: 1, weight: 1}, {index: 2, weight: 31}})
+	ipEntry1 := ipv4Entry(nhgIPv4EntryMap[1], nonDefaultVRF, 1, defaultVRF)
+
+	gRIBI.Modify().AddEntry(t, nh1, nh2, nhg1, ipEntry1)
+
+	if err := awaitTimeout(ctx, gRIBI, t, time.Minute); err != nil {
+		t.Fatalf("Could not program entries via gRIBI, got err: %v", err)
+	}
+
+	// Validate entries were installed in FIB.
+	for _, route := range nhgIPv4EntryMap {
+		chk.HasResult(t, gRIBI.Results(t),
+			fluent.OperationResult().
+				WithIPv4Operation(route).
+				WithOperationType(constants.Add).
+				WithProgrammingResult(fluent.InstalledInFIB).
+				AsResult(),
+			chk.IgnoreOperationID(),
+		)
+	}
+
+	wantWeights := map[string]float64{
+		"1": 1.171,
+		"2": 1.953,
+		"3": 0.402,
+	}
+	// 6.432 weight for vlans 4 to 18.
+	for i := 4; i <= 18; i++ {
+		wantWeights[strconv.Itoa(i)] = 6.432
+	}
+	t.Run("testTraffic", func(t *testing.T) {
+		got := testTraffic(t, ate, top)
+
+		if deviations.HierarchicalWeightResolutionTolerance(dut) != tolerance {
+			tolerance = deviations.HierarchicalWeightResolutionTolerance(dut)
+		}
+		if diff := cmp.Diff(wantWeights, got, cmpopts.EquateApprox(0, tolerance)); diff != "" {
+			t.Errorf("Packet distribution ratios -want,+got:\n%s", diff)
+		}
+	})
+
+	t.Run("validateAFTWeights", func(t *testing.T) {
+		for nhg, weights := range map[uint64][]uint64{
+			2: {3, 5},
+			3: {1, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16, 16},
+		} {
+			got := aftNextHopWeights(t, dut, nhg, defaultVRF)
+			ok := cmp.Equal(weights, got, cmpopts.SortSlices(func(a, b uint64) bool { return a < b }))
+			if !ok {
+				t.Errorf("Valid weights not present for NI: %s, NHG: %d, got: %v, want: %v", defaultVRF, nhg, got, weights)
+			}
+		}
+	})
+
+	// Flush gRIBI routes after test.
+	if err := gribi.FlushAll(gRIBI); err != nil {
+		t.Error(err)
+	}
+}
+
+func TestHierarchicalWeightResolution(t *testing.T) {
+	dut := ondatra.DUT(t, "dut")
+	ate := ondatra.ATE(t, "ate")
+	ctx := context.Background()
+
+	// Configure ATE ports and start Ethernet+IPv4.
+	top := gosnappi.NewConfig()
+	atePort1.configureATE(t, top, ate)
+	atePort2.configureATE(t, top, ate)
+
+	ate.OTG().PushConfig(t, top)
+
+	// configure DUT.
+	configureDUT(t, dut)
+
+	ate.OTG().StartProtocols(t)
```

Comment:
Is this call needed?

Reply:
I think so because it follows the pushconfig

Reply:
We already have to call StartProtocols after pushing traffic config, so this should not be necessary (and I've tested successfully without.)